{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Input file name definitions\n",
    "fileComps = 'Compartments.txt' # input: previous model file\n",
    "fileSpecies = 'Species.txt' # input: previous model file\n",
    "fileRatelaws = 'RatelawsNoSM.txt' # input: previous model file\n",
    "fileOmicsdata = 'OmicsData.txt' # input: previous model file\n",
    "fileGeneReg = 'GeneReg.txt' # input: previous model file\n",
    "\n",
    "# New network and data file name definitions\n",
    "fileMOBILE = 'IRF1_PDL1sub.txt' # input: MOBILE subnetwork\n",
    "fileSPARCEDomics = 'Supplementary_Data_22.xlsx' # input: 10A mRNA & protein levels\n",
    "fileLINCSRNAs = 'RNAseqDataLINCS.csv' # input: LINCS mRNA levels\n",
    "fileLitkvals = 'Schwanhausser2011.txt' # input: Schwanhausser et al. Nature 2011 paper data on mRNA/protein half-lives \n",
    "\n",
    "# Names for the updated input files\n",
    "fileComps_Upd = 'Compartments_MM.txt' # Output: Updated file\n",
    "fileSpecies_Upd = 'Species_MM.txt' # Output: Updated file\n",
    "fileRatelaws_Upd = 'RatelawsNoSM_MM.txt' # Output: Updated file\n",
    "fileOmicsdata_Upd = 'OmicsData_MM.txt' # Output: Updated file\n",
    "fileGeneReg_Upd = 'GeneReg_MM.txt' # Output: Updated file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write file names into 'FileNames' file to inform 'createModel' notebook\n",
    "inputfiles = [fileComps,fileSpecies,fileRatelaws,fileOmicsdata,fileGeneReg]\n",
    "outfiles = [fileComps_Upd,fileSpecies_Upd,fileRatelaws_Upd,fileOmicsdata_Upd,fileGeneReg_Upd]\n",
    "\n",
    "fileNames = open('FileNames.txt','w') \n",
    "for name in inputfiles:\n",
    "    fileNames.write(name+\" \\t\")\n",
    "fileNames.write(\"\\n\")\n",
    "for name in outfiles:\n",
    "    fileNames.write(name+\" \\t\")\n",
    "fileNames.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import required packages and scripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "sys.path.append(os.getcwd()[0:os.getcwd().rfind('/')]+'/bin')\n",
    "\n",
    "import libsbml\n",
    "import importlib\n",
    "import amici\n",
    "import numpy as np\n",
    "import re\n",
    "import pandas as pd\n",
    "from antimony import *\n",
    "from modules.copyDir import copyDirectory\n",
    "\n",
    "# Optional packages to import\n",
    "import amici.plotting\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#copy input files over to current directory if needed\n",
    "current_dir = os.getcwd()\n",
    "input_data_folder = current_dir[0:current_dir.rfind('/')+1]+'input_files'\n",
    "copyDirectory(input_data_folder, os.getcwd()+\"/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read-in Species list\n",
    "network_sheet = np.array([np.array(line.strip().split(\"\\t\")) for line in open(fileMOBILE, encoding='latin-1')])\n",
    "network_sheet = network_sheet[1:]\n",
    "ns = pd.DataFrame(network_sheet)\n",
    "\n",
    "# Read-in TARs structure\n",
    "TARsRead = pd.read_csv(fileGeneReg,header=0,index_col=0,sep=\"\\t\")\n",
    "TARs0 = (TARsRead.values)\n",
    "numberofTARs = len(TARsRead.columns)\n",
    "numberofgenes = len(TARsRead)\n",
    "\n",
    "netGeneNames = np.unique(np.append(network_sheet[:,3],network_sheet[:,6]))\n",
    "netmRNAs = ['m_' + x for x in netGeneNames] # naming convention from SPARCED\n",
    "netProteins = []\n",
    "netRxns_phos = []\n",
    "netGenes = []\n",
    "netTARs = []\n",
    "netSigns = []\n",
    "for idx,val in enumerate(network_sheet[:,8]):    \n",
    "    netProteins = np.append(netProteins,['prot_' + network_sheet[idx,3]])\n",
    "    netProteins = np.append(netProteins,['prot_' + network_sheet[idx,6]])\n",
    "    if val=='2': # flag for protein or chromatin region (2=protein, 4=chromatin)\n",
    "        if \"_\" in network_sheet[idx,7]: # '_' means the RPPA antibody target was a phosphoprotein \n",
    "            pp = network_sheet[idx,7].split(\"_\") # there can be multiple phospho sites\n",
    "            ppprot = netProteins[-1]\n",
    "            for idy,j in enumerate(pp[1:]):\n",
    "                netRxns_phos = np.append(netRxns_phos,ppprot + ' ; ' + ppprot+'_'+j) # Add the rxn \"X-->pX\" or \"pX-->ppX\"\n",
    "                netRxns_phos = np.append(netRxns_phos,ppprot+'_'+j + ' ; ' + ppprot) # Add the rxn \"pX-->X\" or \"ppX-->pX\"\n",
    "                ppprot = ppprot+'_'+j\n",
    "                netProteins = np.append(netProteins,ppprot) # Add the phosphoprotein form in the list of species\n",
    "    # Add the gene-gene connections as TARs for both genes\n",
    "    netGenes = np.append(netGenes,network_sheet[idx,6]) # Add to TARed-gene list\n",
    "    netTARs = np.append(netTARs,['prot_'+network_sheet[idx,3]]) # Add to TARer-protein list\n",
    "    netGenes = np.append(netGenes,network_sheet[idx,3])\n",
    "    netTARs = np.append(netTARs,['prot_'+network_sheet[idx,6]])\n",
    "    # Also assign signs (Activator=1, or repressor=-1) to the TARs\n",
    "    tempS1 = network_sheet[idx,20] # FULL model lasso coefficient column index - from MOBILE\n",
    "    tempS2 = network_sheet[idx,30] # IFNG-LOGO model lasso coefficient column index - from MOBILE\n",
    "    if tempS1>tempS2:\n",
    "        temps = np.sign(np.double(network_sheet[idx,19]))\n",
    "    else:\n",
    "        temps = np.sign(np.double(network_sheet[idx,29]))\n",
    "    netSigns = np.append(netSigns,temps)\n",
    "    netSigns = np.append(netSigns,temps)    \n",
    "netProteins = np.unique(netProteins)\n",
    "\n",
    "TARsReadE = TARsRead.copy()\n",
    "### Translation and nascent protein degradation rxns \n",
    "netRxnsTLs = []\n",
    "netRLawsTLs = []\n",
    "netRxnsTLds = []\n",
    "for idx,gene in enumerate(netGeneNames): \n",
    "    if not gene in TARsReadE.index:\n",
    "        netRxnsTLs = np.append(netRxnsTLs,' ; ' + 'prot_'+ gene) # translation rxn for each new gene\n",
    "        netRLawsTLs = np.append(netRLawsTLs,'kTLn' + str(idx+1) + '*m_'+ gene) # the ratelaw for above\n",
    "        netRxnsTLds = np.append(netRxnsTLds,'prot_'+ gene + ' ; ') # degradation rate for each new nascent protein\n",
    "    else:\n",
    "        print('Warning: Gene \"' + gene + '\" is already in the model. Check and harmonize protein name and rxns!')\n",
    "        netProteins = netProteins[netProteins != ['prot_' + gene]] # remove the protein of this gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# replace the name of the existing genes with active forms of proteins that are actually the TARs\n",
    "netTARs = list(map(lambda x: x.replace('prot_STAT1','STAT1snuc_STAT1snuc'), netTARs)) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Data10A = pd.ExcelFile(fileSPARCEDomics)\n",
    "RNAdata = pd.read_excel(Data10A, 'RNAseq')\n",
    "RNAdata.columns = RNAdata.columns.str.strip()\n",
    "PROTdata = pd.read_excel(Data10A, 'ProteomicsMS')\n",
    "PROTdata.columns = PROTdata.columns.str.strip()\n",
    "OmicsData_old = pd.read_csv(fileOmicsdata,header=0,index_col=0,sep=\"\\t\")\n",
    "\n",
    "Literature_kvals = pd.read_csv(fileLitkvals,header=0,sep=\"\\t\")\n",
    "\n",
    "species_sheet = pd.read_csv(fileSpecies,header=0,index_col=0,sep=\"\\t\",encoding='latin-1')\n",
    "mRNAdf = species_sheet[species_sheet.index.map(lambda s: s.startswith('m_'))] # Separate mRNA species from protein/complexes\n",
    "spsdf = species_sheet[~species_sheet.index.str.startswith(\"m_\")]\n",
    "\n",
    "comps_sheet = pd.read_csv(fileComps,header=0,index_col=0,sep=\"\\t\",encoding='latin-1')\n",
    "# Find the conversion factor from mpc to nM in nucleus\n",
    "Vn = comps_sheet[comps_sheet.index.str.startswith('Nucleus')]['volume'][0]\n",
    "mpc2nmcf_Vn = 1.0E9/(Vn*6.023E+23)\n",
    "\n",
    "OmicsData = OmicsData_old.copy()\n",
    "# First, set protein levels from 10A data, if present\n",
    "for sps in netProteins:    \n",
    "    if sps not in spsdf.index:\n",
    "        pp = sps.split(\"_\")\n",
    "        sps_gene = pp[1]        \n",
    "        tempDF = PROTdata.loc[PROTdata['Gene names'] == sps_gene] # check if the gene has data\n",
    "        if not tempDF.empty: # if yes\n",
    "            aa = tempDF.index.values.astype(int)[0]\n",
    "            bb = (PROTdata.at[aa,'average'])*mpc2nmcf_Vn # Here we define all new species are in nucleus (they are TARs)\n",
    "        else:\n",
    "            bb = 0.0\n",
    "        spsdf.at[sps,'Compartment'] = 'Nucleus'\n",
    "        spsdf.at[sps,'IC'] = str(bb)\n",
    "        spsdf.at[sps,'HGNC'] = sps_gene\n",
    "    else:\n",
    "        print('Species is already in the model, no changes made: ' + sps)\n",
    "\n",
    "for mrna in netmRNAs:    \n",
    "    if mrna not in mRNAdf.index:        \n",
    "        pp = mrna.split(\"_\")\n",
    "        gene = pp[1]        \n",
    "        tempDF = RNAdata.loc[RNAdata['Gene'] == gene] # check if the gene has data\n",
    "        if not tempDF.empty: # if yes\n",
    "            aa = tempDF.index.values.astype(int)[0]\n",
    "            # Below we define all new mRNAs are in nucleus\n",
    "            bbmpc = np.int(np.rint(np.mean([RNAdata.at[aa,'P21-mpc'],RNAdata.at[aa,'P22-mpc']])))\n",
    "            bb = bbmpc*mpc2nmcf_Vn \n",
    "        else:\n",
    "            bb = 0.0\n",
    "        mRNAdf.at[mrna,'Compartment'] = 'Nucleus'\n",
    "        mRNAdf.at[mrna,'IC'] = str(bb) \n",
    "        mRNAdf.at[mrna,'HGNC'] = gene \n",
    "        if gene not in OmicsData.index:\n",
    "            OmicsData.loc[gene] = OmicsData.loc['IGF1R'] # Insert a new row, values the same as for your favorite gene (eg.IGF1R)\n",
    "            OmicsData.at[gene,'Exp RNA'] = str(bbmpc) # set to nearest integer from the mean of two experiments\n",
    "            OmicsData.at[gene,'comp'] = 'Nucleus'\n",
    "    else:\n",
    "        print('mRNA is already in the model, no changes made: ' + mrna)  \n",
    "species_updated = pd.concat([spsdf, mRNAdf], axis=0) # create new species dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Get LINCS datasets\n",
    "RNADataLINCS = pd.read_csv(fileLINCSRNAs,header=0,index_col=0)\n",
    "RNADataLINCS = RNADataLINCS[~RNADataLINCS.index.isnull()] # remove rows without an HGNC identifier\n",
    "\n",
    "mrnaData_IFNG = []\n",
    "for dgene in netGeneNames:\n",
    "    datagenes = RNADataLINCS[RNADataLINCS.index.str.contains(dgene)] #RNAseqData\n",
    "    for gname in datagenes.index.str.split(' '):\n",
    "        for indgene in gname:\n",
    "            if indgene == (dgene):\n",
    "                mrnaData_IFNG.append([indgene, 2**datagenes.at[str(gname[0]),'ctrl_0'],  # take 2^ to linearize the data\n",
    "                                               2**datagenes.at[str(gname[0]),'IFNG_24'],\n",
    "                                               2**datagenes.at[str(gname[0]),'IFNG_48']])\n",
    "            else:\n",
    "                print('boo')\n",
    "mrnaData_t0r2IRF1 = [row[1] for row in mrnaData_IFNG]\n",
    "for ii in np.arange(np.shape(mrnaData_IFNG)[0]): # normalize to m_IRF1 at time=0\n",
    "    mrnaData_t0r2IRF1[ii] = np.float(mrnaData_IFNG[ii][1])/np.float(mrnaData_IFNG[6][1])\n",
    "\n",
    "# get protein/mRNA ratio for IRF1 and use the ratio to calculate new protein levels for MOBILE genes\n",
    "prot_IRF1_10A = species_updated.loc['prot_IRF1'][\"IC\"]\n",
    "mrna_IRF1_10A = species_updated.loc['m_IRF1'][\"IC\"]\n",
    "pr2mr = prot_IRF1_10A/mrna_IRF1_10A # protein/mRNA ratio of IRF1 (the gene that had \"data\" points in both 10A & LINCS datasets)\n",
    "\n",
    "for idx,gene in enumerate(netGeneNames):\n",
    "    if gene!='STAT1': # STAT1 was already present in the model (learned above!)\n",
    "        sps_prot = 'prot_'+gene\n",
    "        sps_mrna = 'm_'+gene\n",
    "    else: # Comment out if do not want to alter STAT1 levels\n",
    "        sps_prot = gene # STAT1 protein was already in the model as \"STAT1\"\n",
    "        sps_mrna = 'm_'+gene\n",
    "\n",
    "    #### Indent below part if the \"else\" is OFF above \n",
    "    # Update mRNA and protein initial levels with LINCS data (note that IRF1 is kept at 10A data levels and \n",
    "    # the rest are calculated based on IRF1 ratios)\n",
    "    species_updated.at[sps_mrna,'IC'] = str(np.float(species_updated.loc['m_IRF1'][\"IC\"])*np.float(mrnaData_t0r2IRF1[idx]))\n",
    "    species_updated.at[sps_prot,'IC'] = str(np.float(species_updated.loc[sps_mrna][\"IC\"])*pr2mr)\n",
    "\n",
    "    # Update the Omics Data input with new mRNA mpc and other parameter values\n",
    "    OmicsData.at[gene,'Exp RNA'] = str(np.int(np.rint(np.float(species_updated.loc[sps_mrna][\"IC\"])/mpc2nmcf_Vn)))\n",
    "    tau_m = 0\n",
    "    if gene!='IRF1': # IRF1 gene name had unrelated genes show up in the dataset, so it is filtered here. \n",
    "        tempGdf = Literature_kvals.loc[Literature_kvals['Gene Names'].str.contains(gene, case=False, na=False)]\n",
    "        # Check for isoform data and use if the gene is not present\n",
    "        tempGdfzz = Literature_kvals.loc[Literature_kvals['Gene Names'].str.contains(gene[0:-1]+'.*', case=False, na=False, regex = True)]    \n",
    "        if not tempGdf.empty:\n",
    "            tau_m = np.float(np.mean(tempGdf['mRNA half-life average [h]']))\n",
    "        elif not tempGdfzz.empty: # Update with mean isoform data if present \n",
    "            tau_m = np.float(np.mean(tempGdfzz['mRNA half-life average [h]']))\n",
    "    if not np.isnan(tau_m) and tau_m!=0:\n",
    "        OmicsData.at[gene,'kTCd'] = str(np.log(2)/tau_m/3600.0)\n",
    "    OmicsData.at[gene,'kTCleak'] = str((np.float(OmicsData.loc[[gene],'kTCd'])*np.float(OmicsData.loc[[gene],'Exp RNA']))*(np.float(OmicsData.loc[[gene],'kGin'])+np.float(OmicsData.loc[[gene],'kGac']))/(np.float(OmicsData.loc[[gene],'kGac'])*np.float(OmicsData.loc[[gene],'Exp GCN'])))\n",
    "\n",
    "### Prepare the new GeneReg structure\n",
    "def_nA = 4.0 # Default value, can change with parameter fitting\n",
    "for gene in netGeneNames: # Add rows for GeneReg\n",
    "    if gene not in TARsReadE.index:\n",
    "        TARsReadE.loc[gene] = str(0)\n",
    "for tempP in netTARs: # Add columns to GeneReg\n",
    "    if tempP not in TARsReadE.columns:\n",
    "        TARsReadE[tempP] = str(0)\n",
    "for idx,gene in enumerate(netGenes): # Set the parameter value pairs for new TARs\n",
    "    if netTARs[idx]=='STAT1snuc_STAT1snuc':\n",
    "        tempProt = 'STAT1' # STAT1snuc=0 at time=0 so use STAT1 total protein level to set KA\n",
    "    else:\n",
    "        tempProt = netTARs[idx]\n",
    "    tempKA = np.double(np.float(species_updated.loc[tempProt][\"IC\"]))\n",
    "    TARsReadE.at[gene,netTARs[idx]] = str(str(netSigns[idx]*def_nA)+'; '+str(np.ceil(tempKA)/2.0))        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Now convert new rxns as row inserts to the model Ratelaws input file\n",
    "# Read-in Ratelaws list\n",
    "ratelaw_sheetE = pd.read_csv(fileRatelaws,header=0,index_col=0,sep=\"\\t\")\n",
    "# Phosphorylation / dephosphorylation rxns -> row insert\n",
    "for idx,val in enumerate(netRxns_phos): \n",
    "    ratelaw_sheetE.at['vMM'+str(idx+1),'Comp_correction'] = 'Nucleus'\n",
    "    ratelaw_sheetE.at['vMM'+str(idx+1),'Species'] = val\n",
    "    ratelaw_sheetE.at['vMM'+str(idx+1),'Ratelaw'] = str(1.0) # This is user defined. \n",
    "# Translation rxns -> row insert\n",
    "for idx,val in enumerate(netRxnsTLs): \n",
    "    ratelaw_sheetE.at['vMMTL'+str(idx+1),'Comp_correction'] = 'Nucleus'\n",
    "    ratelaw_sheetE.at['vMMTL'+str(idx+1),'Species'] = val\n",
    "    ratelaw_sheetE.at['vMMTL'+str(idx+1),'Ratelaw'] = netRLawsTLs[idx]\n",
    "    # If literature data exist for protein half-life, set the kTL accordingly \n",
    "    pp = val.split(\"_\")\n",
    "    gene = pp[1]            \n",
    "    sps_prot = 'prot_'+gene\n",
    "    sps_mrna = 'm_'+gene\n",
    "    tau_p = 0\n",
    "    if gene!='IRF1': \n",
    "        tempGdf = Literature_kvals.loc[Literature_kvals['Gene Names'].str.contains(gene, case=False, na=False)]\n",
    "        # Check for isoform data and use if the gene is not present\n",
    "        tempGdfzz = Literature_kvals.loc[Literature_kvals['Gene Names'].str.contains(gene[0:-1]+'.*', case=False, na=False, regex = True)]    \n",
    "        if not tempGdf.empty:\n",
    "            tau_p = np.float(tempGdf['Protein half-life average [h]'])\n",
    "        elif not tempGdfzz.empty: # Update with mean isoform data if present \n",
    "            tau_p = np.float(np.mean(tempGdfzz['Protein half-life average [h]']))\n",
    "    #         ratelaw_sheetE.at['vMMTL'+str(idx+1),'Unnamed: 4'] = str(np.float(species_updated.loc[sps_prot]['IC'])*(np.log(2)/tau_p/3600.0)/np.float(species_updated.loc[sps_mrna]['IC']))\n",
    "    if not np.isnan(tau_p) and tau_p != 0:\n",
    "        ratelaw_sheetE.at['vMMTL'+str(idx+1),'Unnamed: 4'] = str(np.float(species_updated.loc[sps_prot]['IC'])*(np.log(2)/tau_p/3600.0)/np.float(species_updated.loc[sps_mrna]['IC']))    \n",
    "    else: # if no literature data exist, set to median for SPARCED genes\n",
    "        ratelaw_sheetE.at['vMMTL'+str(idx+1),'Unnamed: 4'] = str(7.122388e-02) # Median value, used for SPARCED-I-SOCS1 model\n",
    "    \n",
    "# Degradation rxns -> row insert\n",
    "for idx,val in enumerate(netRxnsTLds): \n",
    "    ratelaw_sheetE.at['vMMTLd'+str(idx+1),'Comp_correction'] = 'Nucleus'\n",
    "    ratelaw_sheetE.at['vMMTLd'+str(idx+1),'Species'] = val\n",
    "     # If literature data exist for protein half-life, set the kTLd accordingly \n",
    "    pp = val.split(\"_\")\n",
    "    gene = pp[1][0:-3] # to disregard the \" ; \"\n",
    "    sps_prot = 'prot_'+gene\n",
    "    sps_mrna = 'm_'+gene\n",
    "    tau_p = 0\n",
    "    if gene!='IRF1':\n",
    "        tempGdf = Literature_kvals.loc[Literature_kvals['Gene Names'].str.contains(gene, case=False, na=False)]\n",
    "        # Check for isoform data and use if the gene is not present\n",
    "        tempGdfzz = Literature_kvals.loc[Literature_kvals['Gene Names'].str.contains(gene[0:-1]+'.*', case=False, na=False, regex = True)]    \n",
    "        if not tempGdf.empty:\n",
    "            tau_p = np.float(tempGdf['Protein half-life average [h]'])\n",
    "    #         ratelaw_sheetE.at['vMMTLd'+str(idx+1),'Ratelaw'] = str((np.log(2)/tau_p/3600.0))\n",
    "        elif not tempGdfzz.empty: # Update with mean isoform data if present \n",
    "            tau_p = np.float(np.mean(tempGdfzz['Protein half-life average [h]']))\n",
    "    #         ratelaw_sheetE.at['vMMTLd'+str(idx+1),'Ratelaw'] = str((np.log(2)/tau_p/3600.0))\n",
    "    if not np.isnan(tau_p) and tau_p != 0: \n",
    "        ratelaw_sheetE.at['vMMTLd'+str(idx+1),'Ratelaw'] = str((np.log(2)/tau_p/3600.0))\n",
    "    else: # if no literature data exist, calculate or set to median value (4.599338e-06) for SPARCED genes\n",
    "        ratelaw_sheetE.at['vMMTLd'+str(idx+1),'Ratelaw'] = str(np.double(ratelaw_sheetE.loc['vMMTL'+str(idx+1),'Unnamed: 4'])*np.double(species_updated.loc[sps_mrna,'IC'])/np.double(species_updated.loc[sps_prot,'IC']))       \n",
    "# USER CHECK: Here, also set the STAT1 degradation rate so that it is constant without IFNg stimulation\n",
    "gene = 'STAT1'\n",
    "sps_prot = gene\n",
    "sps_mrna = 'm_'+gene\n",
    "ratelaw_sheetE.at['vI61','Ratelaw'] = str(np.double(ratelaw_sheetE.loc['vI50','Unnamed: 4'])*np.double(species_updated.loc[sps_mrna,'IC'])/np.double(species_updated.loc[sps_prot,'IC']))       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "comps_sheet.to_csv(fileComps_Upd,sep='\\t',header=True, index=True)\n",
    "species_updated.to_csv(fileSpecies_Upd,sep='\\t',header=True, index=True)\n",
    "ratelaw_sheetE.to_csv(fileRatelaws_Upd,sep='\\t',header=True, index=True)\n",
    "OmicsData.to_csv(fileOmicsdata_Upd,sep='\\t',header=True, index=True)\n",
    "TARsReadE.to_csv(fileGeneReg_Upd,sep='\\t',header=True, index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "netGeneNames = pd.DataFrame(netGeneNames)\n",
    "netGeneNames.to_csv('NetGeneNames.txt',sep='\\t',header=True, index=True)"
   ]
  }
 ],
 "metadata": {
  "file_extension": ".py",
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  },
  "mimetype": "text/x-python",
  "name": "python",
  "npconvert_exporter": "python",
  "pygments_lexer": "ipython3",
  "version": 3
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
