{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Model Simulation\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is how to simulate an existing SPARCED model, consisting of an SBML file and scripts for random sampling of mRNA counts for stochastic single-cell simulations. Below is an illustration of how a single-cell is simulated, with required inputs and arguments."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## First, import the required packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import libsbml\n",
    "import importlib\n",
    "import amici\n",
    "import amici.plotting\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "from datetime import datetime\n",
    "import scipy.stats\n",
    "import argparse\n",
    "import sys\n",
    "import os\n",
    "sys.path.append(os.getcwd()[0:os.getcwd().rfind('/')]+'/bin')\n",
    "from modules.RunSPARCED import RunSPARCED\n",
    "%matplotlib inline\n",
    "\n",
    "import time\n",
    "import random"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Input file names\n",
    "fileSPARCEDomics = 'Supplementary_Data_22.xlsx' # input: 10A mRNA & protein levels\n",
    "fileLINCSRNAs = 'RNAseqData.csv' # input: LINCS mRNA levels\n",
    "fileLINCSProts = 'RPPADataLINCS.txt' # input: LINCS mRNA levels\n",
    "fileLINCSProtsStd = 'RPPADataStdLINCSfc.txt' # input: LINCS mRNA levels\n",
    "\n",
    "### Read-in data\n",
    "# Proteomics\n",
    "RPPAData = pd.read_csv(fileLINCSProts,header=0,index_col=1,sep=\"\\t\")\n",
    "RPPAData = RPPAData[~RPPAData.index.isnull()] # remove rows without an HGNC identifier\n",
    "RPPADataSTD = pd.read_csv(fileLINCSProtsStd,header=0,index_col=1,sep=\"\\t\")\n",
    "RPPADataSTD = RPPADataSTD[~RPPADataSTD.index.isnull()] # remove rows without an HGNC identifier\n",
    "# Transcriptomics\n",
    "RNAseqData = pd.read_csv(fileLINCSRNAs,header=0,index_col=0)\n",
    "RNAseqData = RNAseqData[~RNAseqData.index.isnull()] # remove rows without an HGNC identifier"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Access proteomic data for the two proteins added via MEMMAL\n",
    "spslist = ['prot_IRF1','prot_CD274']\n",
    "dataEGF = []\n",
    "dataEIfn = []\n",
    "dataEGFstdfc = []\n",
    "dataEIfnstdfc = []\n",
    "for idx,sps in enumerate(spslist):\n",
    "    pp = sps.split(\"_\")\n",
    "    sps_gene = pp[1]        \n",
    "    tempDF = RPPAData.loc[RPPAData.index == sps_gene] # check if the gene has data\n",
    "    dataEGF.append((2.0)**np.array(tempDF[['CTRL_0','EGF_1','EGF_4','EGF_8','EGF_24','EGF_48']]))\n",
    "    dataEIfn.append((2.0)**np.array(tempDF[['CTRL_0','IFNG_1','IFNG_4','IFNG_8','IFNG_24','IFNG_48']]))\n",
    "    tempDFs = RPPADataSTD.loc[RPPADataSTD.index == sps_gene] # check if the gene has data\n",
    "    dataEGFstdfc.append((2.0)**np.array(tempDFs[['CTRL_0','EGF_1','EGF_4','EGF_8','EGF_24','EGF_48']]))\n",
    "    dataEIfnstdfc.append((2.0)**np.array(tempDFs[['CTRL_0','IFNG_1','IFNG_4','IFNG_8','IFNG_24','IFNG_48']]))   \n",
    "dataEGF = np.array(dataEGF)\n",
    "dataEIfn = np.array(dataEIfn)\n",
    "dataEGF = dataEGF.reshape(2,6)\n",
    "dataEIfn = dataEIfn.reshape(2,6)\n",
    "dataEGFstdfc = np.array(dataEGFstdfc)\n",
    "dataEIfnstdfc = np.array(dataEIfnstdfc)\n",
    "dataEGFstdfc = dataEGFstdfc.reshape(2,6)\n",
    "dataEIfnstdfc = dataEIfnstdfc.reshape(2,6)\n",
    "dataEGFfc = dataEGF.copy()\n",
    "for idx,row in enumerate(dataEGF):\n",
    "    dataEGFfc[idx,:] = row/row[0]\n",
    "dataEIfnfc = dataEIfn.copy()\n",
    "for idx,row in enumerate(dataEIfnfc):\n",
    "    dataEIfnfc[idx,:] = row/row[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Access rna data for the six genes added via MEMMAL\n",
    "spslistm = ['m_ACSL5','m_BST2','m_CLIC2','m_FAM83D','m_HIST2H2AA3','m_METAP2']\n",
    "mdataEGF = []\n",
    "mdataEIfn = []\n",
    "for idx,sps in enumerate(spslistm):\n",
    "    pp = sps.split(\"_\")\n",
    "    sps_gene = pp[1]        \n",
    "    tempDF = RNAseqData.loc[RNAseqData.index == sps_gene] # check if the gene has data\n",
    "    mdataEGF.append((2.0)**np.array(tempDF[['CTRL_0','EGF_24','EGF_48']]))\n",
    "    mdataEIfn.append((2.0)**np.array(tempDF[['CTRL_0','IFNG_24','IFNG_48']]))\n",
    "mdataEGF = np.array(mdataEGF)\n",
    "mdataEIfn = np.array(mdataEIfn)\n",
    "mdataEGF = mdataEGF.reshape(6,3)\n",
    "mdataEIfn = mdataEIfn.reshape(6,3)\n",
    "mdataEGFfc = mdataEGF.copy()\n",
    "for idx,row in enumerate(mdataEGF):\n",
    "    mdataEGFfc[idx,:] = row/row[0]\n",
    "mdataEIfnfc = mdataEIfn.copy()\n",
    "for idx,row in enumerate(mdataEIfnfc):\n",
    "    mdataEIfnfc[idx,:] = row/row[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot the data we have\n",
    "tPoints = [0.0,1.0,2.0,3.0,4.0,5.0]\n",
    "tPoints = np.reshape(tPoints,[1,6])\n",
    "\n",
    "for idx,sps in enumerate(spslist):    \n",
    "    fig = plt.figure(figsize=(4, 1), dpi=600, facecolor='w', edgecolor='k') # dpi=300,         \n",
    "    ax1 = fig.add_subplot(121)\n",
    "    barlist1 = plt.bar(tPoints.flatten(), np.reshape(dataEGF[idx,:],[1,6]).flatten(), color='darkorange')\n",
    "    plt.errorbar(tPoints.flatten(), np.reshape(dataEGF[idx,:],[1,6]).flatten(), \n",
    "                 yerr = np.reshape(dataEGFstdfc[idx,:],[1,6])[0], fmt='none', elinewidth=1.0,\n",
    "                 capsize=1.5, zorder=2, ecolor='black')\n",
    "    barlist1[0].set_color('dimgrey')\n",
    "    plt.ylim([0,5])\n",
    "    plt.xticks(ticks=[0,1,2,3,4,5], labels=[])\n",
    "    \n",
    "    ax2 = fig.add_subplot(122)\n",
    "    barlist2 = plt.bar(tPoints.flatten(), np.reshape(dataEIfn[idx,:],[1,6]).flatten(), color='firebrick')\n",
    "    plt.errorbar(tPoints.flatten(), np.reshape(dataEIfn[idx,:],[1,6]).flatten(),\n",
    "                 yerr = np.reshape(dataEIfnstdfc[idx,:],[1,6])[0], fmt='none', elinewidth=1.0,\n",
    "                 capsize=1.5, zorder=2, ecolor='black')\n",
    "    barlist2[0].set_color('dimgrey')\n",
    "    plt.ylim([0,5])\n",
    "    plt.xticks(ticks=[0,1,2,3,4,5], labels=[])\n",
    "    plt.yticks(ticks=[0,2.5,5], labels=[])\n",
    "#     plt.savefig(sps+'_bar.svg', dpi=600, format='svg') # edgecolor='auto', backend=None\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Define the model SBML file name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "### Run simulations to see the parameter fitting to LINCS data\n",
    "# The mRNA species we have data for, will plot their trajectories against data\n",
    "spslistm = ['m_ACSL5','m_BST2','m_CLIC2','m_FAM83D','m_HIST2H2AA3','m_METAP2']\n",
    "# The protein species we have data for, will plot their trajectories against data\n",
    "spslist = ['prot_IRF1','prot_CD274']\n",
    "\n",
    "# Input file names\n",
    "fileComps = 'Compartments_MM.txt' \n",
    "fileSpecies = 'Species_MM.txt' \n",
    "fileRatelaws = 'RatelawsNoSM_MM.txt' \n",
    "fileOmicsdata = 'OmicsData_MM.txt' \n",
    "fileGeneReg = 'GeneReg_MM.txt'\n",
    "\n",
    "sbml_file = 'MEMMAL_o4a.xml' # SBML model file\n",
    "model_name = sbml_file[0:-4]\n",
    "model_output_dir = model_name\n",
    "        \n",
    "tPoints = np.array([0.0,1.0, 4.0, 8.0, 24.0, 48.0])*3600.0/60.0 # LINCS RPPA data points\n",
    "tPoints = tPoints.astype(int)\n",
    "tPoints = tPoints.reshape(1,6)\n",
    "       \n",
    "parnames0 = [# IRF1 regulations\n",
    "             'k4059_5','k4064_5','k4065_5',          \n",
    "             'k4067_5','k4068_5','k4069_5',\n",
    "             'k4070_5',\n",
    "             # PD-L1 regulations\n",
    "             'k4064_3','k4065_3','k4067_3',          \n",
    "             'k4068_3','k4069_3','k4071_5',\n",
    "             # The rest of new TARs\n",
    "             'k4059_3',\n",
    "             'k4066_3','k4066_5','k4066_7','k4066_9','k4066_11','k4066_13',\n",
    "             'k4070_3','k4070_7','k4070_9','k4070_11','k4070_13','k4070_15',\n",
    "             'k4071_3',\n",
    "             # kTLs\n",
    "             'k4064_1',\n",
    "             'k4065_1',\n",
    "             'k4067_1',\n",
    "             'k4225', # kTCd_FAM83D\n",
    "             'k4068_1',\n",
    "             'k4069_1',\n",
    "             'k4071_1',\n",
    "             'k4066_1',\n",
    "             'k4070_1']\n",
    "paravals0 = [50.0,5.5,2.5,\n",
    "             5.8,1.0,5.8,\n",
    "             50.0,\n",
    "             150.0,50.0,150.0,\n",
    "             50.0,19.0,150.0,\n",
    "             8.0,\n",
    "             32.0,2.0,1.0,0.5,10.0,10.0,\n",
    "             1.7,50.0,50.0,50.0,150.0,1.0,\n",
    "             200.0,\n",
    "             0.00042768,\n",
    "             0.010040722,\n",
    "             0.00126,\n",
    "             0.00122,\n",
    "             0.10754,\n",
    "             0.0062905,\n",
    "             0.02217178,\n",
    "             0.0027585905,\n",
    "             0.01100385]\n",
    "\n",
    "### Run to steady state with new parameter values\n",
    "sys.path.insert(0, os.path.abspath(model_output_dir))\n",
    "model_module0 = importlib.import_module(model_name)\n",
    "model0 = model_module0.getModel()\n",
    "solver0 = model0.getSolver() # Create solver instance\n",
    "solver0.setMaxSteps = 1e50 \n",
    "spICs0 = np.array(model0.getInitialStates())\n",
    "sps_list0 = model0.getStateIds()\n",
    "th = 1000\n",
    "model0.setFixedParameterById('k4064_2',0.1)\n",
    "model0.setFixedParameterById('k4065_2',0.1)\n",
    "model0.setFixedParameterById('k4066_2',0.1)\n",
    "model0.setFixedParameterById('k4067_2',0.1)\n",
    "model0.setFixedParameterById('k4068_2',0.1)\n",
    "model0.setFixedParameterById('k4069_2',0.1)\n",
    "model0.setFixedParameterById('k4070_2',0.1)\n",
    "model0.setFixedParameterById('k4071_2',0.1)\n",
    "for idx,param in enumerate(paravals0):\n",
    "    model0.setFixedParameterById(parnames0[idx],param)\n",
    "model0.setInitialStates(spICs0)\n",
    "model0.setTimepoints(np.linspace(0,th*3600,1001)) # set timepoints ,NSteps+1\n",
    "rdata_S1MMstst = amici.runAmiciSimulation(model0, solver0) \n",
    "\n",
    "### From st.st, run to 24 hrs with IGF only\n",
    "spICs1 = np.array(rdata_S1MMstst['x'][-1,:])\n",
    "model_module1 = importlib.import_module(model_name)\n",
    "model1 = model_module1.getModel()\n",
    "solver1 = model1.getSolver() # Create solver instance\n",
    "solver1.setMaxSteps = 1e50 \n",
    "sps_list1 = model1.getStateIds()\n",
    "th = 24\n",
    "IGFdose = 0.02\n",
    "EGFdose = 1.5625\n",
    "IFNgDose = 1.1834\n",
    "spICs1[160] = IGFdose # IGF \n",
    "model1.setFixedParameterById('k4064_2',0.1)\n",
    "model1.setFixedParameterById('k4065_2',0.1)\n",
    "model1.setFixedParameterById('k4066_2',0.1)\n",
    "model1.setFixedParameterById('k4067_2',0.1)\n",
    "model1.setFixedParameterById('k4068_2',0.1)\n",
    "model1.setFixedParameterById('k4069_2',0.1)\n",
    "model1.setFixedParameterById('k4070_2',0.1)\n",
    "model1.setFixedParameterById('k4071_2',0.1)\n",
    "for idx,param in enumerate(paravals0):\n",
    "    model1.setFixedParameterById(parnames0[idx],param)\n",
    "model1.setInitialStates(spICs1)\n",
    "model1.setTimepoints(np.linspace(0,th*3600,1001)) # set timepoints ,NSteps+1\n",
    "rdata_S1MM_IGF24 = amici.runAmiciSimulation(model1, solver1) \n",
    "\n",
    "### From IGF 24 hr, run to 48 hrs with EGF only\n",
    "spICs2 = np.array(rdata_S1MM_IGF24['x'][-1,:])\n",
    "model_module2 = importlib.import_module(model_name)\n",
    "model2 = model_module2.getModel()\n",
    "solver2 = model2.getSolver() # Create solver instance\n",
    "solver2.setMaxSteps = 1e50 \n",
    "sps_list2 = model2.getStateIds()\n",
    "th = 48\n",
    "IGFdose = 0.02\n",
    "EGFdose = 1.5625\n",
    "IFNgDose = 1.1834\n",
    "spICs2[160] = IGFdose # IGF \n",
    "spICs2[155] = EGFdose # EGF \n",
    "model2.setFixedParameterById('k4064_2',0.1)\n",
    "model2.setFixedParameterById('k4065_2',0.1)\n",
    "model2.setFixedParameterById('k4066_2',0.1)\n",
    "model2.setFixedParameterById('k4067_2',0.1)\n",
    "model2.setFixedParameterById('k4068_2',0.1)\n",
    "model2.setFixedParameterById('k4069_2',0.1)\n",
    "model2.setFixedParameterById('k4070_2',0.1)\n",
    "model2.setFixedParameterById('k4071_2',0.1)\n",
    "for idx,param in enumerate(paravals0):\n",
    "    model2.setFixedParameterById(parnames0[idx],param)\n",
    "model2.setInitialStates(spICs2)\n",
    "model2.setTimepoints(np.linspace(0,th*3600,2881)) # set timepoints ,NSteps+1\n",
    "rdata_S1MM_48hrE = amici.runAmiciSimulation(model2, solver2) \n",
    "\n",
    "### From st.st, run to 48 hrs with EGF+IFNG\n",
    "spICs3 = np.array(rdata_S1MM_IGF24['x'][-1,:])\n",
    "model_module3 = importlib.import_module(model_name)\n",
    "model3 = model_module3.getModel()\n",
    "solver3 = model3.getSolver() # Create solver instance\n",
    "solver3.setMaxSteps = 1e50 \n",
    "sps_list3 = model3.getStateIds()\n",
    "n_sp = len(sps_list3)\n",
    "th = 48\n",
    "IGFdose = 0.02\n",
    "EGFdose = 1.5625\n",
    "IFNgDose = 1.1834\n",
    "spICs3[160] = IGFdose # IGF \n",
    "spICs3[155] = EGFdose # EGF \n",
    "spICs3[801] = IFNgDose # IFNg \n",
    "model3.setFixedParameterById('k4064_2',0.1)\n",
    "model3.setFixedParameterById('k4065_2',0.1)\n",
    "model3.setFixedParameterById('k4066_2',0.1)\n",
    "model3.setFixedParameterById('k4067_2',0.1)\n",
    "model3.setFixedParameterById('k4068_2',0.1)\n",
    "model3.setFixedParameterById('k4069_2',0.1)\n",
    "model3.setFixedParameterById('k4070_2',0.1)\n",
    "model3.setFixedParameterById('k4071_2',0.1)\n",
    "for idx,param in enumerate(paravals0):\n",
    "    model3.setFixedParameterById(parnames0[idx],param)\n",
    "model3.setInitialStates(spICs3)\n",
    "model3.setTimepoints(np.linspace(0,th*3600,2881)) # set timepoints ,NSteps+1\n",
    "rdata_S1MM_48hrEIfn = amici.runAmiciSimulation(model3, solver3)\n",
    "\n",
    "pltList = (['m_ACSL5','prot_ACSL5',\n",
    "            'm_BST2','prot_BST2',\n",
    "            'm_CLIC2','prot_CLIC2',\n",
    "            'm_FAM83D','prot_FAM83D',\n",
    "            'm_HIST2H2AA3','prot_HIST2H2AA3',\n",
    "            'm_METAP2','prot_METAP2', \n",
    "            'm_STAT1','STAT1', 'STAT1snuc_STAT1snuc',\n",
    "            'm_SOCS1','SOCS1', \n",
    "            'm_IRF1','prot_IRF1',\n",
    "            'm_CD274','prot_CD274'])\n",
    "\n",
    "for idx,sps in enumerate(pltList):    \n",
    "    fig = plt.figure(figsize=(6, 1), dpi=100, facecolor='w', edgecolor='k')\n",
    "    ax1 = fig.add_subplot(121)\n",
    "    plt.plot(rdata_S1MMstst['t']/3600.0, rdata_S1MMstst['x'][:,sps_list1.index(sps)], 'k')\n",
    "    ax1.set_ylabel(sps)\n",
    "    plt.xlim([-100,(rdata_S1MMstst['t'][-1]/3600.0)+1]) \n",
    "    ax2 = fig.add_subplot(122)\n",
    "    plt.plot(rdata_S1MM_48hrE['t']/3600.0, rdata_S1MM_48hrE['x'][:,sps_list2.index(sps)], 'k')\n",
    "    plt.plot(rdata_S1MM_48hrEIfn['t']/3600.0, rdata_S1MM_48hrEIfn['x'][:,sps_list3.index(sps)], 'm:')\n",
    "    plt.xlim([-1,th+1])\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Make sure MEMMAL expansion did not mess-up SPARCED-IFNG model responses\n",
    "spslistplt = ['STAT1snuc_STAT1snuc','SOCS1','m_SOCS1']\n",
    "ylimMaxs = [30,8,15,3,5,2]\n",
    "for idx,sps in enumerate(spslistplt):  \n",
    "    fig = plt.figure(figsize=(6, 1), dpi=600, facecolor='w', edgecolor='k') # dpi=300,         \n",
    "    ax1 = fig.add_subplot(231)\n",
    "    aaa = (rdata_S1MM_48hrEIfn['x'][:,sps_list3.index(sps)]).flatten()\n",
    "    bbb = aaa/aaa.max()\n",
    "    plt.plot(rdata_S1MM_48hrEIfn['t']/3600.0, bbb, 'dimgrey', linewidth=1.2)\n",
    "    plt.xlim([-1,th+1])\n",
    "    plt.ylim([-0.1,1.2])   \n",
    "    plt.xticks(np.arange(0, th+1, step=8)) \n",
    "#     plt.savefig(sps+'.svg', dpi=600, format='svg') # edgecolor='auto', backend=None\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "### Plot protein and mRNA levels for the species we have data for\n",
    "tPoints = [0.0,1.0,4.0,8.0,24.0,48.0]\n",
    "tPoints = np.reshape(tPoints,[1,6])\n",
    "\n",
    "mtPoints = [0.0,24.0,48.0]\n",
    "mtPoints = np.reshape(mtPoints,[1,3])\n",
    "\n",
    "for idx,sps in enumerate(spslist):    \n",
    "    fig = plt.figure(figsize=(6, 1), dpi=600, facecolor='w', edgecolor='k') # dpi=300,         \n",
    "    ax1 = fig.add_subplot(121)\n",
    "    plt.errorbar(tPoints.flatten(), np.reshape(dataEGFfc[idx,:],[1,6]).flatten(), \n",
    "                 yerr = np.reshape(dataEGFstdfc[idx,:],[1,6])[0], fmt='none', elinewidth=1.5,\n",
    "                 capsize=2, zorder=2, ecolor='darkorange')\n",
    "    aaa = rdata_S1MM_48hrE['x'][:,sps_list2.index(sps)]\n",
    "    plt.plot(rdata_S1MM_48hrE['t']/3600.0, aaa/aaa[0] , 'dimgrey', zorder=1) \n",
    "\n",
    "    plt.errorbar(tPoints.flatten(), np.reshape(dataEIfnfc[idx,:],[1,6]).flatten(),\n",
    "                 yerr = np.reshape(dataEIfnstdfc[idx,:],[1,6])[0], fmt='none', elinewidth=1.5,\n",
    "                 capsize=2, zorder=2, ecolor='firebrick')\n",
    "    bbb = rdata_S1MM_48hrEIfn['x'][:,sps_list3.index(sps)]\n",
    "    plt.plot(rdata_S1MM_48hrEIfn['t']/3600.0, bbb/bbb[0], 'k', zorder=1)\n",
    "    plt.xlim([-1,th+1])\n",
    "    plt.ylim([0,6])\n",
    "    plt.xticks(np.arange(0, th+1, step=8))  \n",
    "    ax1.set_ylabel(sps+' (nM)')\n",
    "    ax1.set_xlabel('Time (hr)')\n",
    "    ax1.ticklabel_format(useOffset=False)\n",
    "#     plt.savefig(sps+'.svg', dpi=600, format='svg') # edgecolor='auto', backend=None\n",
    "    plt.show()\n",
    "\n",
    "ylimMaxs = [30,8,15,3,5,2]\n",
    "for idx,sps in enumerate(spslistm):  \n",
    "    fig = plt.figure(figsize=(6, 1), dpi=600, facecolor='w', edgecolor='k') # dpi=300,         \n",
    "    ax1 = fig.add_subplot(121)\n",
    "    plt.plot(mtPoints, np.reshape(mdataEGFfc[idx,:],[1,3]), 'darkorange', marker='o', markersize=2.5, zorder=2)\n",
    "    aaa = (rdata_S1MM_48hrE['x'][:,sps_list2.index(sps)]).flatten()\n",
    "    bbb = aaa/aaa[0]\n",
    "    plt.plot(rdata_S1MM_48hrE['t']/3600.0, bbb, 'dimgrey', zorder=1)\n",
    "\n",
    "    plt.plot(mtPoints, np.reshape(mdataEIfnfc[idx,:],[1,3]), 'firebrick', marker='x', markersize=2.5, zorder=2)\n",
    "    ccc = (rdata_S1MM_48hrEIfn['x'][:,sps_list3.index(sps)]).flatten()\n",
    "    ddd = ccc/ccc[0]\n",
    "    plt.plot(rdata_S1MM_48hrEIfn['t']/3600.0, ddd, 'k', zorder=1)\n",
    "    plt.xticks(np.arange(0, th+1, step=8))  \n",
    "    ax1.set_ylabel(sps+' (nM)')\n",
    "    ax1.set_xlabel('Time (hr)')\n",
    "    ax1.ticklabel_format(useOffset=False)\n",
    "    plt.xlim([-1,th+1])\n",
    "    plt.ylim([-0.3,ylimMaxs[idx]])   \n",
    "#     plt.savefig(sps+'.svg', dpi=600, format='svg') # edgecolor='auto', backend=None\n",
    "    plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
